### PR TITLE
Use brave-core debounce service and settings in iOS

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/36033

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
See issue for STR
Simulator Screenshot - iPhone 14 - 2024-02-07 at 11 23 01

### CODE REVIEW

**Comment by cuba** [comment]
> Changes
> ### Security Hotspots
> 1. **Debouncing Service Implementation (`DebounceService.swift`)**: The introduction of a debouncing service for URL processing could present security risks if not properly validated, particularly in parsing and redirect logic. Ensure URL validation and rule processing are securely implemented to prevent potential manipulation.

The debounce service is a very dumb wrapper which delegates all the functionality to the brave-core (tried and tested) brave-core components.

> 2. **De-AMP Preferences Handling (`deAmpPrefs`)**: While not extensively detailed in the provided diff, the manner in which De-AMP preferences (`deAmpPrefs`) are stored, updated, and accessed could pose security concerns, particularly if user preferences are not securely managed or if the feature can be externally manipulated.

Using the existing prefs used already on desktop.

> 3. **Migration Logic for De-AMP and Debouncing (`Migration.swift`)**: The migration logic adds complexity and potential risk, especially if the migration process fails to securely transition user preferences, leading to unintended behavior or security vulnerabilities related to user settings.

The migration is pretty straightforward and is called in a place where other migrations are already happening.

File: ios/brave-ios/Sources/Brave/Frontend/Browser/Tab.swift [hunk 1]
```diff
@@ -391,7 +391,7 @@ class Tab: NSObject {
     $0.addTarget(self, action: #selector(reload), for: .valueChanged)
   }
 
-  func createWebview() {
+  func createWebview(includeDeAmpScript: Bool) {
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/Tab.swift:394]: This doesn't seem like an ideal solution

Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/Tab.swift:394]: This is the first time we use a setting that controls a script that comes from brave core:

```
    let scriptPreferences: [UserScriptManager.ScriptType: Bool] = [
        .cookieBlocking: Preferences.Privacy.blockAllCookies.value,
        .mediaBackgroundPlay: Preferences.General.mediaAutoBackgrounding.value,
        .nightMode: Preferences.General.nightModeEnabled.value,
        .deAmp: includeDeAmpScript,
        .playlistMediaSource: Preferences.Playlist.webMediaSourceCompatibility.value,
      ]
```

So far we've been using global prefs but the braveCore is not global. Maybe I need to change `DeAmpPrefs` to a `DeAmpService` and create a factory for it? Then I can access it from the tab as I wish

File: ios/brave-ios/Sources/Brave/Migration/Migration.swift [hunk 2]
```diff
@@ -44,6 +47,21 @@ public class Migration {
     )
   }
 
+  private func deAmpMigration() {
+    if let isDeAmpEnabled = Preferences.Shields.autoRedirectAMPPagesDeprecated.value {
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Brave/Migration/Migration.swift:51]: `guard let` if you're going to have this in a separate function

File: ios/brave-ios/Sources/Brave/Migration/Migration.swift [hunk 3]
```diff
@@ -44,6 +47,21 @@ public class Migration {
...
+      braveCore.deAmpPrefs.isDeAmpEnabled = isDeAmpEnabled
+      Preferences.Shields.autoRedirectAMPPagesDeprecated.value = nil
+    }
+  }
+
+  private func debounceMigration() {
+    if let isDebounceEnabled = Preferences.Shields.autoRedirectTrackingURLsDeprecated.value {
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Brave/Migration/Migration.swift:58]: `guard let` if you're going to have this in a separate function

File: ios/brave-ios/Sources/Brave/Migration/Migration.swift [hunk 4]
```diff
@@ -35,6 +35,9 @@ public class Migration {
       Preferences.Playlist.firstLoadAutoPlay.value = true
     }
 
+    deAmpMigration()
+    debounceMigration()
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Brave/Migration/Migration.swift:39]: `migrateDeAmpPreferences`, `migrateDebouncePreferences`

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**